### DISCUSSION TITLE
Link libdevice into a single file.

### DISCUSSION THREAD

**Discussion by bader** [intel/llvm#7301] [Ideas]
I'd like to discuss the idea of merging `libdevice` functionality into a single file.
https://github.com/intel/llvm/pull/6999 adds a new file, DPC++ compiler driver will unbundle and link to the user's device code. 
We are about to have 20 different files and as I can see we always link all 20 of them. 
Multi-file distribution is valuable when user has controls over which files are linked. I'm not sure if we have such option.

Tagging @jinge90, @mdtoguchi.

**Comment by mdtoguchi** [comment]
We do have an option which gives us some control over the device libs that are pulled in named `-fno-sycl-device-lib=<arg>` where we have control at 'group' levels but not giving control over specific libraries individually.

**Reply by bader** [reply]
Thanks! I think it might be reasonable to disable linking "fallback" implementations when there will be extensions implemented in GPU drivers.
Today, I don't know if `-fno-sycl-device-lib=<arg>` is useful. I see that libraries are linked with `-only-needed` flag, which means that unused files are dropped automatically. If some library is excluded via `-fno-sycl-device-lib` option, the output file will have unresolved external symbol, which GPU driver won't resolve, and compilation will fail. Right?
What is the use case for this option? 
My search on GitHub returned only usages in lit tests in intel/llvm forks.

**Comment by mdtoguchi** [comment]
> If some library is excluded via -fno-sycl-device-lib option, the output file will have unresolved external symbol, which GPU driver won't resolve, and compilation will fail. Right?

Yes, that's correct.

> What is the use case for this option?

Perhaps somebody wanted to provide their own or what is being provided has proven to be problematic - but I may just be reaching for some usefulness.
